### DISCUSSION TITLE
how do I mock node_module packages with jest under Nx?

### DISCUSSION THREAD

**Discussion by gregveres** [nrwl/nx#10382] [General]
Can someone point out where I need to place the __mocks__ directory for mocking out node_module packages?
I have tried in the workspace directory but that doesn't seem to get picked up. 
Do I have to replicate the __mock__ dir in every lib and app directory? 

I have searched and searched but can't find any guidance on this.

Thanks

**Comment by gregveres** [comment]
A big thanks to Nathan Thibert on slack for answering my question. I will put the answer here in case someone else looks for the solution in the future. 

There are two parts to the answer:

1. you place the __mocks__ directory at the root of your workspace: ```myOrg/__mocks__```. This should be right beside the top level node_modules directory
2. You need to modify your jest config to add a root entry. This can be done in your jest.preset.js file, also in the workspace root folder. Here is the one that I am using:
```
const nxPreset = require('@nrwl/jest/preset').default;
const path = require('path');

module.exports = {
  ...nxPreset,
  roots: ["<rootDir>", path.resolve(__dirname, "./__mocks__")]
 };
```

**Reply by rfprod** [reply]
@gregveres if you want to unit test your mocks, there is another way. Generate a unit-testing library, put the testing code there and import it where needed. See an example below.

- [a library test-setup.ts](https://github.com/rfprod/nx-ng-starter/blob/main/libs/client-core-components/src/test-setup.ts)
- [the unit testing library jest globals patch](https://github.com/rfprod/nx-ng-starter/blob/443b121057513c1e861cc8a072ee3bfe79540317/libs/client-unit-testing/src/lib/jest/patch/jest-jsdom-globals.patch.ts#L18)
- [the unit testing library jest external packages mock](https://github.com/rfprod/nx-ng-starter/blob/443b121057513c1e861cc8a072ee3bfe79540317/libs/client-unit-testing/src/lib/jest/external/external.mock.ts#L3)

With this approach you can unit test your mocks as well because it's part of a workspace library.

**Comment by gregveres** [comment]
@rfprod Thanks. That is an interesting approach that I hadn't thought of. 
I always wrap my use of Document or Window, so it is easy for me to mock the library that provides the wrappers, but I do see what you are doing for the external libraries. I do like that approach. 

I am on the fence between having all the mocks in the one unit test library. I prefer to have the mock closer to the code that is being mocked - sort of like having the testing code next to the code that is being tested. 

Thanks for pointing out your approach. I learned a new technique.

**Reply by rfprod** [reply]
> I am on the fence between having all the mocks in the one unit test library. I prefer to have the mock closer to the code that is being mocked - sort of like having the testing code next to the code that is being tested.

Actually, it makes sense to put very specific testing code closer to the code that's being tested, but it depends on the context.
At the same time general purpose testing code can go to a shared unit testing library.

For instance, I had to sacrifice being dry to make a publishable library self-contained https://github.com/rfprod/nx-ng-starter/blob/main/libs/client-d3-charts/src/test-setup.ts I have same d3 mocks in two places in the workspace.

I am still considering whether I should consume the d3 mocks in the unit testing library from the d3-charts library.
In my case it is possible because the d3-charts library does not depend on the unit-testing library. If it were, and the unit testing library consumed mocks from the d3-charts lib, there would be a circular dependency.

**Comment by barrymichaeldoyle** [comment]
Thank you so much for this! It saved me hours of digging around!
